app/graphql/concerns/fancy_mutation.rb

```diff
@@ -0,0 +1,196 @@
+    def errors(*types)
+      errors_union.possible_types(*types)
+    end
+
+    # Adds types to the union of warning types for the mutation
+    # @param types [Array<Class>] The warning types this can return
+    def warnings(*types)
+      warnings_union.possible_types(*types)
+    end
+
+    ### private_class_methods ###
+
+    # These methods are kinda strange. The first time they're called, they create a new type, hook
+    # it up to the mutation, and return it. When they get called again, they return that same type,
+    # allowing us to modify them repeatedly from different locations. This allows utilities to add
+    # types automatically such as an ignore_warnings argument, or a common error type.
+    def errors_union
+      @errors_union ||= begin
+        union_name = "#{graphql_name}ErrorsUnion"
+        errors_union = Class.new(Types::Union::Base) do
+          graphql_name union_name
+
+          def self.resolve_type(object, _context)
+            object.key?(:__type) ? object[:__type] : super
+          end
+        end
+        field :errors, [errors_union], null: true
+
+        errors_union
+      end
+    end
+
+    def warnings_union
+      @warnings_union ||= begin
+        union_name = "#{graphql_name}WarningsUnion"
+        warnings_union = Class.new(Types::Union::Base) do
+          graphql_name union_name
+
+          def self.resolve_type(object, _context)
+            object.key?(:__type) ? object[:__type] : super
+          end
+        end
+        field :warnings, [warnings_union], null: true
+
+        # Add the ignore_warnings argument and error type
+        input { argument :ignore_warnings, GraphQL::Types::Boolean, required: false }
+        errors Types::Errors::WarningsPresent
+
+        warnings_union
+      end
+    end
+
+    def input_type
+      @input_type ||= begin
+        input_name = "#{graphql_name}Input"
+        input_type = Class.new(Types::Input::Base) do
+          graphql_name input_name
+        end
+        argument :input, input_type, required: true
+
+        input_type
+      end
+    end
+  end
+
+  included do
+    prepend PrependedMethods
+    private_class_method :errors_union
+    private_class_method :warnings_union
+    private_class_method :input_type
+  end
+
+  # Adds an error if there is no current user session.
+  def authenticate!
+    errors << Types::Errors::NotAuthenticated.build if current_user.blank?
+    true
+  end
+
+  # Adds an error if the current user is not authorized to perform the action on the object.
+  # @param object [ActiveRecord::Base] The object to check authorization on
+  # @param action [Symbol] The action to check authorization for
+  def authorize!(object, action, policy: Pundit::PolicyFinder.new(object).policy)
+    authorized = policy.new(current_token, object).public_send(action)
+    errors << Types::Errors::NotAuthorized.build(object: object, action: action) unless authorized
+    true
+  end
```

Comment:
These two methods are helpers for use in `ready?` and `authorized?` methods, to make it easy